src/common/transformations/src/transformations/common_optimizations/StateManagementPattern.cpp

```diff
@@ -0,0 +1,230 @@
+        q,
+        std::make_shared<pattern::op::Or>(OutputVector{k_concat, k_shaped, k_shaped_transposed, k_simply_shaped}),
+        std::make_shared<pattern::op::Or>(OutputVector{v_concat, v_shaped, v_shaped_transposed, v_simply_shaped}),
+        std::make_shared<pattern::op::Or>(OutputVector{sdpa_mask, pattern::any_input()})
+    });
+
+    ov::matcher_pass_callback callback = [OV_CAPTURE_CPY_AND_THIS, &kv_parameters, &model_remaining_params, &sliding_window, &parameters_to_remove, &assignes_to_remove](ov::pass::pattern::Matcher& m) {
+        std::cout << "____" << matcher_name << "___Matched___" << std::endl;
+        OPENVINO_ASSERT(m.get_pattern_value_map().find(sdpa) != m.get_pattern_value_map().end());
+        const auto& pattern_map = m.get_pattern_map();
+        OPENVINO_ASSERT(pattern_map.find(sdpa) != pattern_map.end());
+        auto real_q = pattern_map.at(q);
+        // Why do have double check here?
+
+        // takes option that has 4D instead of fine-grained Reshape analysis
+        // it avoids complication in the pattern, but we don't really have many options
+        auto take_4d = [OV_CAPTURE_CPY_AND_THIS](std::shared_ptr<Node> option1, std::shared_ptr<Node> option2, std::shared_ptr<Node> option3) {
+            // Question: should it be get_partial_shape() of output or input ???
+            if (pattern_map.find(option1) != pattern_map.end() && pattern_map.at(option1)->input(0).get_partial_shape().rank().get_length() == 4) {
+                return pattern_map.at(option1);
+            } else if (pattern_map.at(option2)->input(0).get_partial_shape().rank().get_length() == 4) {
+                return pattern_map.at(option2);
+            } else {
+                return pattern_map.at(option3);
+            }
+        };
+
+        auto real_k = take_4d(k_current, k_current_reshaped, k_current2);
+        auto real_v = take_4d(v_current, v_current_reshaped, v_current2);
+
+        // is_cpu() check required here.
+        // kv_cache_type required here (CURRENTLY USING STUB)
+        auto kv_cache_type = element::i64;
+        auto k_parameter = std::make_shared<v0::Parameter>(kv_cache_type, PartialShape{-1, -1, -1, -1});
+        auto v_parameter = std::make_shared<v0::Parameter>(kv_cache_type, PartialShape{-1, -1, -1, -1});
+        kv_parameters.push_back(k_parameter);
+        kv_parameters.push_back(v_parameter);
+        auto kv_transpose_order = v0::Constant::create(element::i32, Shape{} ,{0, 2, 1, 3}); //correct Constant creation? What Shape to use?
+        auto q_transpose = std::make_shared<v1::Transpose>(real_q, kv_transpose_order);
+        auto q_reshape = std::make_shared<v1::Reshape>(q_transpose, v0::Constant::create(element::i32, Shape{1}, {0, 0, -1}), true); //correct Constant creation? What Shape to use?
+
+        std::shared_ptr<Node> k_transpose_order = kv_transpose_order; // eeeh, is it a right way to assign Constants? Maybe I should clone somehow?
```

Comment:
TODO: is it a right way to assign Constants?

Reply:
That's OK, it is just a shared pointer to the node. In the code here and below we are building the correct transposition order that have a few options. There is a default order that is applied when we are migrating data from main graph flow to PA and back. But also there may be additional transposes in some models that we need to mix with the default one. In this case we are building pure-constant sub-expression based on that default order constant node. This subgraph will be folded with CF later. That's a common approach in transformations -- do not compute values in the code, let CF do that for you. We are not modifying this default `kv_transpose_order` so we reuse the same constant node in many places in the graph, no need for cloning.

Reply:
Ok, leaving as it is.